src/combinators.rs

```diff
@@ -0,0 +1,396 @@
+    let mut results = Vec::new();
+    while !input.is_empty() && is_backtrace(condition(input))? {
+      match combinator(input) {
+        Ok((result_input, value)) => {
+          results.push(value);
+          input = result_input;
+        }
+        Err(ParseError::Backtrace) => {
+          return Ok((input, results));
+        }
+        Err(err) => return Err(err),
+      }
+    }
+    Ok((input, results))
+  }
+}
+
+/// Keeps consuming a combinator into an array until a condition
+/// is met or backtracing occurs.
+pub fn separated_list<'a, O, OSeparator>(
+  combinator: impl Fn(&'a str) -> ParseResult<'a, O>,
+  separator: impl Fn(&'a str) -> ParseResult<'a, OSeparator>,
+) -> impl Fn(&'a str) -> ParseResult<'a, Vec<O>> {
+  move |mut input| {
+    let mut results = Vec::new();
+    while !input.is_empty() {
+      match combinator(input) {
+        Ok((result_input, value)) => {
+          results.push(value);
+          input = result_input;
+        }
+        Err(ParseError::Backtrace) => {
+          return Ok((input, results));
+        }
+        Err(err) => return Err(err),
+      }
+      input = match separator(input) {
+        Ok((input, _)) => input,
+        Err(ParseError::Backtrace) => break,
+        Err(err) => return Err(err),
+      };
+    }
+    Ok((input, results))
+  }
+}
+
+/// Applies the parser 0 or more times and returns a vector
+/// of all the parsed results.
+pub fn many0<'a, O>(
+  combinator: impl Fn(&'a str) -> ParseResult<'a, O>,
+) -> impl Fn(&'a str) -> ParseResult<'a, Vec<O>> {
+  many_till(combinator, |_| ParseError::backtrace::<()>())
+}
+
+/// Skips the whitespace.
+pub fn skip_whitespace(input: &str) -> ParseResult<()> {
+  for (pos, c) in input.char_indices() {
+    if !c.is_whitespace() {
+      return Ok((&input[pos..], ()));
+    }
+  }
+
+  Ok(("", ()))
+}
+
+/// Checks if a condition is true for a combinator.
+pub fn if_true<'a, O>(
+  combinator: impl Fn(&'a str) -> ParseResult<'a, O>,
+  condition: impl Fn(&O) -> bool,
+) -> impl Fn(&'a str) -> ParseResult<'a, O> {
+  move |input| {
+    let (input, value) = combinator(input)?;
+    if condition(&value) {
+      Ok((input, value))
+    } else {
+      ParseError::backtrace()
+    }
+  }
+}
+
+/// Checks if a combinator is true without consuming the input.
+#[allow(dead_code)]
+pub fn check<'a, O>(
+  combinator: impl Fn(&'a str) -> ParseResult<'a, O>,
+) -> impl Fn(&'a str) -> ParseResult<'a, ()> {
+  move |input| match combinator(input) {
+    Ok(_) => Ok((input, ())),
+    Err(_) => ParseError::backtrace(),
+  }
+}
+
+/// Checks if a combinator is false without consuming the input.
+pub fn check_not<'a, O>(
+  combinator: impl Fn(&'a str) -> ParseResult<'a, O>,
+) -> impl Fn(&'a str) -> ParseResult<'a, ()> {
+  move |input| match combinator(input) {
+    Ok(_) => ParseError::backtrace(),
+    Err(_) => Ok((input, ())),
+  }
+}
+
+/// Logs the result for quick debugging purposes.
+#[cfg(debug_assertions)]
+#[allow(dead_code)]
```

Comment:
Remove?

Reply:
I want to keep this one because it's useful for quick debugging of the parser. I have it as only being compiled in debug.